src/features/devices/models/device-additions.ts

```diff
@@ -4,8 +4,139 @@ import type {
 	AndNode,
 	BooleanTypeNodes,
 	EqualsNode,
+	OrNode,
 } from '@balena/abstract-sql-compiler';
 
+// The Cloud API has the notion of active/inactive devices via a field
+// on the device resource. Computing the overall status below depends
+// on whether the device 'is active'.
+//
+// We must somehow expose the overall status/progress attributes on the
+// model and we could contrive a way for the cloud to inject the extra
+// cases, but that would quickly become overly complex because the order
+// of cases matters. We'd never be able to cleanly make the core API
+// agnostic to the cloud.
+//
+// Therefore, we opted for using the 'is active' field when it's part of
+// the model, otherwise we use a dummy `false` value everywhere that field
+// is referenced. This approach can also be used in core API
+// translations in a consistent way.
+export const isInactiveFn = (isActiveNotExist: boolean): BooleanTypeNodes =>
+	isActiveNotExist
+		? ['Boolean', false]
+		: ['Not', ['ReferencedField', 'device', 'is active']];
+
+export const isOverallOffline: AndNode = [
+	'And',
+	['Not', ['ReferencedField', 'device', 'is online']],
+	[
+		'In',
+		['ReferencedField', 'device', 'api heartbeat state'],
+		['EmbeddedText', 'offline'],
+		['EmbeddedText', 'unknown'],
+	],
+];
+
+// VPN not connected and supervisor didn't yet reach the API,
+// so it's still provisioning.
+export const isPreProvisioning: AndNode = [
+	'And',
+	['Not', ['ReferencedField', 'device', 'is online']],
+	['NotExists', ['ReferencedField', 'device', 'last connectivity event']],
+	[
+		'Equals',
+		['ReferencedField', 'device', 'api heartbeat state'],
+		['EmbeddedText', 'unknown'],
+	],
+];
+
+export const isPostProvisioning: EqualsNode = [
+	'Equals',
+	['ReferencedField', 'device', 'provisioning state'],
+	['EmbeddedText', 'Post-Provisioning'],
+];
+// This check does not double check heartbeat state as it is already checked from isOverallOffline which runs before
+export const hasPartialConnectivity: OrNode = [
+	'Or',
+	[
+		'And',
+		['Equals', ['ReferencedField', 'device', 'is online'], ['Boolean', true]],
+		[
+			'NotEquals',
+			['ReferencedField', 'device', 'api heartbeat state'],
+			['EmbeddedText', 'online'],
+		],
+	],
+	[
+		'And',
+		['Equals', ['ReferencedField', 'device', 'is online'], ['Boolean', false]],
+		[
+			'In',
+			['ReferencedField', 'device', 'api heartbeat state'],
+			['EmbeddedText', 'online'],
+			['EmbeddedText', 'timeout'],
+		],
```

Comment:
Shouldn't we only check for online heartbeat, so that this matches w/ how we check the heartbeat in the vpn false case?
```suggestion
						'Equals',
						['ReferencedField', 'device', 'api heartbeat state'],
						['EmbeddedText', 'online'],
```
I mean, for a device w/ the vpn disabled, ie a device that we can infer its state only from its heartbeat state, timeout is probably the only `reduced-functionality` case. My reasoning:
* vpn disabled & heartbeat online => operational
* vpn disabled & heartbeat offline/unknown => disconnected
* vpn disabled & heartbeat timeout => reduced-functionality (even this doesn't feel concrete enough, but that's the only that's left)